### PR TITLE
Gubbins - Bacterial recombination dectection

### PR DESCRIPTION
**Pull request by Slugger70**
Tool from the Sanger institute for detecting recombination of bacterial genomes.. Requires a whole genome alignment file in fasta format.

### CODE REVIEW

File: tools/gubbins/gubbins.xml [hunk 1]
```diff
@@ -0,0 +1,235 @@
+<tool id="gubbins" name="Gubbins - Recombination detection in Bacteria" version="0.1.0">
+    <requirements>
+        <requirement type="package" version="2.2.1">gubbins</requirement>
+    </requirements>
+    <stdio>
+        <exit_code range="1:" />
+    </stdio>
```
Comment (shiltemann) [inline tools/gubbins/gubbins.xml]: you can remove the stdio section and use `detect_errors="exit_code"` on the command tag

File: tools/gubbins/gubbins.xml [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    <version_command>run_gubbins.py --version</version_command>
+    <command><![CDATA[
+        
+        cp "$alignment_file" "foo.aln" &&
```
Comment (shiltemann) [inline tools/gubbins/gubbins.xml]: does this file get big? could you replace the copy with a symlink?

File: tools/gubbins/gubbins.xml [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+      
+        run_gubbins.py
+        
+        --threads \${GALAXY_SLOTS:-1}
+        
+        #if $adv.iters
+          -i "$adv.iters"
```
Comment (shiltemann) [inline tools/gubbins/gubbins.xml]: could you use single quotes for all these quoted params in cheetah please?

File: tools/gubbins/gubbins.xml [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+      <test>
+        <param name="alignment_file" value="multiple_recombinations.aln" ftype="fasta" />
+        <param name="conv_meth" value="recombination" />
+        <output name="recomb_pred_gff">
+          <assert_contents>
+            <has_text text="##gff-version 3" />
+          </assert_contents>
+        </output>
+      </test>
+    </tests>
```
Comment (shiltemann) [inline tools/gubbins/gubbins.xml]: there seem to be more files in the test-data directory than are used in the tests? are all the files needed?

File: tools/gubbins/gubbins.xml [hunk 5]
```diff
@@ -0,0 +1,235 @@
+<tool id="gubbins" name="Gubbins - Recombination detection in Bacteria" version="0.1.0">
```
Comment (shiltemann) [inline tools/gubbins/gubbins.xml]: perhaps the `Recombination detection in Bacteria` part could be in the `<description>` tag? looks a little nicer in the toolbox imho

File: tools/gubbins/gubbins.xml [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        <param name="tree_builder" type="select" label="Tree builder" help="Application to use for tree building, default RAxML">
+          <option value="raxml" selected="true">RAxML</option>
+          <option value="fasttree">FastTree</option>
+          <option value="hybrid">Hybrid</option>
+        </param>
+        <param name="min_snps" type="integer" label="Minimum SNPS" value="3" help="Min SNPs to identify a recombination block, default is 3" />
+        <param name="filter_percentage" type="integer" label="Filter Percentage" value="25" help="Filter out taxa with more than this percentage of gaps, default is 25" />
+        <param name="min_window_size" type="integer" label="Minimum Window Size" value="100" help="Minimum window size, default 100" />
+        <param name="max_window_size" type="integer" label="Maximum Window Size" value="10000" help="Maximum window size, default 10000" />
+        <param name="remove_duplicates" type="boolean" label="Remove identical sequences" false_value="" true_value="-d" value="false" />
```
Comment (shiltemann) [inline tools/gubbins/gubbins.xml]: `true_value` and `false_value` should not have the underscore, and use `checked=` not `value=` to indicate the default (but unchecked is the default so can also be left out as well)
https://docs.galaxyproject.org/en/latest/dev/schema.html

File: tools/gubbins/gubbins.xml [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+        
+        #if $really_adv.min_window_size
+          -a $really_adv.min_window_size
+        #end if
+        
+        #if $really_adv.max_window_size
+          -b $really_adv.max_window_size
+        #end if
+        
+        "$remove_duplicates"
```
Comment (shiltemann) [inline tools/gubbins/gubbins.xml]: `$really_adv.remove_duplicates` I think?

File: tools/gubbins/gubbins.xml [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+      <data format="vcf" name="summ_snp_vcf" label="${tool.name} on ${on_string} Summary of SNP Distribution vcf" from_work_dir="foo.summary_of_snp_distributions.vcf" >
+        <filter>outfiles and 'vcf' in outfiles</filter>
+      </data>
+      <data format="embl" name="base_branch_embl" label="${tool.name} on ${on_string} Base Branch Reconstruction embl" from_work_dir="foo.base_branch_reconstruction.embl" >
+        <filter>outfiles and 'baseb' in outfiles</filter>
+      </data>
+    </outputs>
+    
+    
+    <tests>
```
Comment (shiltemann) [inline tools/gubbins/gubbins.xml]: only the gff output is tested, could you add tests or checks for the other outputs as well?

File: tools/gubbins/gubbins.xml (hunk 4 above)
Comment (Slugger70) [inline tools/gubbins/gubbins.xml]: done

File: tools/gubbins/gubbins.xml (hunk 8 above)
Comment (Slugger70) [inline tools/gubbins/gubbins.xml]: done

File: tools/gubbins/gubbins.xml (hunk 6 above)
Comment (Slugger70) [inline tools/gubbins/gubbins.xml]: done

File: tools/gubbins/gubbins.xml (hunk 7 above)
Comment (Slugger70) [inline tools/gubbins/gubbins.xml]: done

File: tools/gubbins/gubbins.xml (hunk 3 above)
Comment (Slugger70) [inline tools/gubbins/gubbins.xml]: done

File: tools/gubbins/gubbins.xml (hunk 2 above)
Comment (Slugger70) [inline tools/gubbins/gubbins.xml]: done

File: tools/gubbins/gubbins.xml (hunk 1 above)
Comment (Slugger70) [inline tools/gubbins/gubbins.xml]: done

File: tools/gubbins/gubbins.xml (hunk 5 above)
Comment (Slugger70) [inline tools/gubbins/gubbins.xml]: done

**Comment by Slugger70** [comment]
Thanks for the review Saskia!

File: tools/gubbins/gubbins.xml (hunk 8 above)
Comment (shiltemann) [inline tools/gubbins/gubbins.xml]: thanks :) any chance we could make one test that checks all the boxes for desired outputs and tests all of them?

**Comment by shiltemann** [comment]
you will also need to make a `.shed.yml` file btw

File: tools/gubbins/gubbins.xml (hunk 8 above)
Comment (Slugger70) [inline tools/gubbins/gubbins.xml]: Done

File: tools/gubbins/gubbins.xml [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+    <description>Recombination detection in Bacteria</description>
+    <requirements>
+        <requirement type="package" version="2.2.1">gubbins</requirement>
+    </requirements>
+    
+    <version_command>run_gubbins.py --version</version_command>
+    
+    <command detect_errors="exit_code"><![CDATA[
+        
+        ln -s "$alignment_file" "foo.aln" &&
```
Comment (shiltemann) [inline tools/gubbins/gubbins.xml]: this input file param still needs some single quotes :)

**Comment by shiltemann** [comment]
looking good @Slugger70! :) ..it seems that the test-data you removed were added back again in the latest commit, could you remove those again? 
and a final little nitpick: could you use 4-space indentation throughout? (it's mixed 2 and 4 now)

**Comment by Slugger70** [comment]
Hi Saskia, the data are the other outputs you asked me to test. That's why they are back in the repo. I will do the indentation thing..

**Comment by shiltemann** [comment]
@Slugger70 if you test with `assert_contents` you do not need the files in the test-data directory (it will just check if the output it created during the test contains that string). If you want to test that the entire file is the same you can use `file="myfile.ext"` instead of `assert_contents`

**Comment by Slugger70** [comment]
Ok, I'll delete them again.

**Comment by shiltemann** [comment]
thanks @Slugger70 !

**Pull request merged**